samples/cellular/nrf_cloud_multi_service/src/provisioning_support.c

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+
+#include <modem/lte_lc.h>
+#include <modem/nrf_modem_lib.h>
+#include <net/nrf_provisioning.h>
+
+#include "cloud_connection.h"
+
+LOG_MODULE_REGISTER(cloud_provisioning, CONFIG_MULTI_SERVICE_LOG_LEVEL);
+
+#define NETWORK_UP			BIT(0)
+#define NETWORK_DOWN			BIT(1)
+
+static K_EVENT_DEFINE(prov_events);
+static bool provisioning_started = false;
+
+/* Called by the provisioning library to request the LTE modem be taken offline before credential
+ * installation, and then again after credential installation is complete. This is necessary because
+ * credentials cannot be installed to the modem while it is online.
+ *
+ * This limitation only applies to the nrf91 modem; Wi-Fi, for instance, is unaffected. But for
+ * simplicity we use conn_mgr to take take all network interfaces offline and online.
+ */
+static int modem_mode_cb(enum lte_lc_func_mode new_mode, void *user_data)
+{
+	enum lte_lc_func_mode fmode;
+
+	ARG_UNUSED(user_data);
+
+	/* The nrf_provisioning library requires us to return the previous functional mode. */
+	if (lte_lc_func_mode_get(&fmode)) {
+		LOG_ERR("Failed to read modem functional mode");
+		return -EFAULT;
+	}
+
+	if (new_mode == LTE_LC_FUNC_MODE_NORMAL || new_mode == LTE_LC_FUNC_MODE_ACTIVATE_LTE) {
+		LOG_DBG("Provisioning library requests normal mode");
+
+		/* Connect all interfaces since we are done provisioning. */
+		conn_mgr_all_if_connect(true);
+
+		/* Wait for network readiness to be re-established before returning. */
+		LOG_DBG("Waiting for network up");
+		k_event_wait(&prov_events, NETWORK_UP, false, K_FOREVER);
+
+		LOG_DBG("Network is up.");
+	} else {
+		LOG_DBG("Provisioning library requests offline mode");
+		/* Disconnect all interfaces to allow cert installation. */
+		conn_mgr_all_if_disconnect(true);
+
+		/* Wait for network disconnection before returning. */
+		LOG_DBG("Waiting for network down");
+		k_event_wait(&prov_events, NETWORK_DOWN, false, K_FOREVER);
+
+		LOG_DBG("Network is down.");
+	}
+
+	return fmode;
+}
+static struct nrf_provisioning_mm_change mmode = { .cb = modem_mode_cb };
+
+/* Initialize the provisioning library and start checking for provisioning commands.
+ * Called automatically the first time network connectivity is established.
+ */
+void start_provisioning(void)
+{
+	await_date_time_known(K_FOREVER);
```

Comment:
any concern with a forever wait in the l4_event_handler?

Reply:
Yeah, good point, I should really have caught that myself. I'll need to rework this handling a bit to avoid that.